allensdk/brain_observatory/behavior/project_apis/data_io/behavior_project_cloud_api.py

```diff
@@ -0,0 +1,290 @@
+import pandas as pd
+from typing import Iterable, Union, Dict, List
+from pathlib import Path
+import logging
+import ast
+import semver
+
+from allensdk.brain_observatory.behavior.project_apis.abcs import (
+    BehaviorProjectBase)
+from allensdk.brain_observatory.behavior.behavior_session import (
+    BehaviorSession)
+from allensdk.brain_observatory.behavior.behavior_ophys_experiment import (
+    BehaviorOphysExperiment)
+from allensdk.api.cloud_cache.cloud_cache import S3CloudCache
+from allensdk import __version__ as sdk_version
+
+
+# [min inclusive, max exclusive)
+COMPATIBILITY = {
+        "pipeline_versions": {
+            "2.9.0": {"AllenSDK": ["2.9.0", "3.0.0"]}}}
+
+
+class BehaviorCloudCacheVersionException(Exception):
+    pass
+
+
+def version_check(pipeline_versions: List[Dict[str, str]],
+                  sdk_version: str = sdk_version,
+                  compatibility: Dict[str, Dict] = COMPATIBILITY):
+    """given a pipeline_versions list (from manifest) determine
+    the pipeline version of AllenSDK used to write the data. Lookup
+    the compatibility limits, and check the the running version of
+    AllenSDK meets those limits.
+
+    Parameters
+    ----------
+    pipeline_versions: List[Dict[str, str]]:
+        each element has keys name, version, (and comment - not used here)
+    sdk_version: str
+        typically the current return value for allensdk.__version__
+    compatibility_dict: Dict
+        keys (under 'pipeline_versions' key) are specific version numbers to
+        match a pipeline version for AllenSDK from the manifest. values
+        specify the min (inclusive) and max (exclusive) limits for
+        interoperability
+
+    Raises
+    ------
+    BehaviorCloudCacheVersionException
+
+    """
+    pipeline_version = [i for i in pipeline_versions
+                        if "AllenSDK" == i["name"]]
+    if len(pipeline_version) != 1:
+        raise BehaviorCloudCacheVersionException(
+                "expected to find 1 and only 1 entry for `AllenSDK` "
+                "in the manifest.data_pipeline metadata. "
+                f"found {len(pipeline_version)}")
+    pipeline_version = pipeline_version[0]["version"]
+    if pipeline_version not in compatibility["pipeline_versions"]:
+        raise BehaviorCloudCacheVersionException(
+                f"no version compatibility listed for {pipeline_version}")
+    version_limits = compatibility["pipeline_versions"][pipeline_version]
+    pver = sdk_version
+    smin = semver.VersionInfo.parse(version_limits["AllenSDK"][0])
+    smax = semver.VersionInfo.parse(version_limits["AllenSDK"][1])
+    if (pver < smin) | (pver >= smax):
+        raise BehaviorCloudCacheVersionException(
+                f"expected {smin} <= {pipeline_version} < {smax}")
```

Comment:
Since this will be end-user (external scientist) facing, I would have much more verbose error messages.

Maybe something like:

```
The version of the visual-behavior-ophys data files (specified in path_to_users_current_release_manifest) requires that your AllenSDK version be between x and y. Your version of AllenSDK is: allensdk.__version__. If you want to use the specified manifest to retrieve data, please upgrade or downgrade AllenSDK to the range specified. If you just want to get the latest version of visual-behavior-ophys data please upgrade to the latest AllenSDK version and try this process again.
```

Reply:
done.